### PR TITLE
Add Devise Configuration

### PR DESCRIPTION
**Pull request by alx1313**
Add the devise configuration based on the instructions of devise,
also generate the database, added the first migration and a home route to verify the conection.

After you pull this branch dont forget to bundle:
```
bundle install
```

Co-authored-by: alx1313 <alx2406@hotmail.com>
Co-authored-by: CarlosSe <carlos.seggch@gmail.com>
Co-authored-by: jorge27 <ja.garciavega@outlook.com>
Co-authored-by: DiegoFigueroa98 <16460315@colima.tecnm.mx>
Co-authored-by: pacheco-andres <andrespach4@gmail.com>
Co-authored-by: psantollo <psantollo@ucol.mx>
Co-authored-by: Ruli-AlSe <ra.almazaserrano@gmail.com>

### CODE REVIEW

**Review by Ruli-AlSe — APPROVED** [review]
Hi guys, all looks good to me!, thanks.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1 @@
+123,125
```
Comment (jaimelr) [inline .vscode/settings.json:1]: I believe this file does not belong to the device configuration. My suggestion is to remove it from the PR.

File: app/controllers/home_controller.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class HomeController < ApplicationController
+  def index
+  end
+end
```
Comment (jaimelr) [inline app/controllers/home_controller.rb:4]: What do you think of adding a new line at the end of the file to avoid this warning on Github. It's kind of a best practice: https://thoughtbot.com/blog/no-newline-at-end-of-file

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
```
Comment (jaimelr) [inline app/models/user.rb:3]: Comments are an important part of the code, but in order for them to be helpful they should be meaningful and provide context and extra information that cannot be directly infered by the code it relates to. In this case, I think comments do not provide too much of a valuable information so maybe we can remove them. What do you think?

File: app/views/layouts/home/index.html.erb [hunk 4]
```diff
@@ -0,0 +1 @@
+<h1>Hello BrightCoders</h1>
```
Comment (jaimelr) [inline app/views/layouts/home/index.html.erb:1]: Please check my last comment about adding a new line at the of the files and let me know if you agree.

**Review by jaimelr — COMMENTED** [review]
Looks good. Just a couple of minor comments, let me know if you have questions about them.

File: config/environments/development.rb [hunk 5]
```diff
@@ -36,6 +36,9 @@
 
   config.action_mailer.perform_caching = false
 
+  # Mailer Devise Configuration
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
```
Comment (softr8) [inline config/environments/development.rb:40]: try as much as possible to not modify config/application.rb or config/initializers/* files, this is mostly due the fact that rake app:upgrade makes a diff between current version and new one when upgrading rails, everything should be done via custom initializer

File: app/views/devise/sessions/new.html.erb [hunk 6]
```diff
@@ -1,26 +1,41 @@
-<h2>Log in</h2>
+<div class="container-fluid p-0 m-0">
+    <div class="image-container">
+        <div class="wrap-image">
+            <div class="form-container">
+                <%= image_tag 'login-img.png', class: "mx-auto d-block img-fluid" %>
+                <hr>
+                  <% flash.each do |type, msg| %>
+                  <div class="alert alert-info">
```
Comment (softr8) [inline app/views/devise/sessions/new.html.erb:8]: type is usually used to change alert color, something like `class="alert alert-#{type}"` and customize background color per message type

File: app/views/devise/sessions/new.html.erb [hunk 7]
```diff
@@ -1,26 +1,41 @@
...
+                    <%= msg %>
+                  </div>
+                <% end %>
+                <h4 class="py-4 text-center">Login to continue</h4>
+                <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
```
Comment (softr8) [inline app/views/devise/sessions/new.html.erb:13]: there's a gem named bootstrap_form_for that wraps all bootstrap forms and make it super easy to create beautiful forms using bootstrap classes

File: app/views/layouts/application.html.erb [hunk 8]
```diff
@@ -10,8 +10,12 @@
   </head>
 
   <body>
-    <p class="notice"><%= notice %></p>
+    <% if !notice.nil? %>
```
Comment (softr8) [inline app/views/layouts/application.html.erb:13]: try to use everything inside flash messages but with custom css per message type
{ notice: 'green', warning: 'yellow', danger: 'red', success: 'green'}

**Review by Ilyeo — CHANGES_REQUESTED** [review]

File: Rakefile [hunk 9]
```diff
@@ -2,6 +2,10 @@
 # for example lib/tasks/capistrano.rake, and they will automatically be available to Rake.
 
 require_relative 'config/application'
+require 'rubocop'
+require 'rubocop/rake_task'
+# require 'rubocop-rake'
+require 'rubocop-rails'
```
Comment (EduardoGHdez) [inline Rakefile:8]: Looks like this is not part of the devise configuration.

I wonder why do we need them 😄 
And if we really want to keep this change. WDYT about moving this to another PR for describing the motivation of this change?

File: app/models/user.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+  devise :database_authenticatable, :registerable,
```
[the remark this answers is not in the document]
Comment (EduardoGHdez) [reply app/models/user.rb:4]: For more info: https://github.com/BrightCoders-Proyectos/control-almacen/blob/master/README-PROJECT.md

File: db/migrate/20200914221259_devise_create_users.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+class DeviseCreateUsers < ActiveRecord::Migration[6.0]
+  def change
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :email,              null: false, default: ""
```
Comment (EduardoGHdez) [inline db/migrate/20200914221259_devise_create_users.rb:7]: Looks like we want to allow users to sign-in by using their employee-number
Any plans for adding that column in this table? 😄 

Can we confirm if the email is needed?
I mean, it'd be useful to have it since it'd help with the reset-password functionality

File: app/views/devise/sessions/new.html.erb (hunk 6 above)
Comment (EduardoGHdez) [reply app/views/devise/sessions/new.html.erb:8]: Sweet!

File: app/views/devise/sessions/new.html.erb (hunk 7 above)
Comment (EduardoGHdez) [reply app/views/devise/sessions/new.html.erb:13]: `bootstrap_form_for` is a really good gem, but I'd avoid adding more dependencies and would encourage early rails learners to use the helpers that rails already provides.

WDYT? 😄

File: spec/login_spec.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
+require 'rails_helper'
```
Comment (EduardoGHdez) [inline spec/login_spec.rb:1]: I'm glad to see acceptance testing in the early stage
Good job! 😄 

I'd just put this specs into a `features` or `system` directory. It's a common practice when using Capybara & RSpec 

https://github.com/teamcapybara/capybara#using-capybara-with-rspec

File: test/fixtures/users.yml [hunk 13]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (dannharo) [inline test/fixtures/users.yml:1]: delete this and the `test/models/user_test.rb` files because you are not using them, you are using `rspec` (the `spec/` directory), WDYT?

File: test/models/user_test.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (dannharo) [inline test/models/user_test.rb:1]: you can delete this file

File: spec/models/admin_spec.rb [hunk 15]
```diff
@@ -0,0 +1 @@
+require 'rails_helper'
```
Comment (dannharo) [inline spec/models/admin_spec.rb:1]: you are not writing any spec here.

File: spec/features/login_spec.rb [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+      admin = FactoryBot.create(:admin)
+      visit new_admin_session_path
+  
+      fill_in 'Email', with: admin.email
+      fill_in 'Password', with: admin.password
+      click_button 'Log in'
+
+      expect(page).to have_content('Signed in successfully.')
+    end
+  end
```
Comment (dannharo) [inline spec/features/login_spec.rb:35]: check the indentation in this file

File: test/models/user_test.rb (hunk 14 above)
Comment (dannharo) [reply test/models/user_test.rb:1]: I think you use the `generate` command to create the `user` model, that creates these files, then you change it to `admin` and now you are using `rspec`

**Review by Ilyeo — APPROVED** [review]

File: app/views/layouts/application.html.erb (hunk 8 above)
Comment (alx1313) [reply app/views/layouts/application.html.erb]: understood, thanks!

File: app/views/devise/sessions/new.html.erb (hunk 6 above)
Comment (CarlosSe) [reply app/views/devise/sessions/new.html.erb]: We implemented a helper function to change the background of alerts, messages look better

File: app/views/devise/sessions/new.html.erb (hunk 7 above)
Comment (CarlosSe) [reply app/views/devise/sessions/new.html.erb:13]: bootstrap_form is listed on README in technology section, we add class: 'form-control' to inputs, it's repetitive, but I think not all team has worked with bootstrap, which option is better in this case?

File: spec/features/login_spec.rb (hunk 16 above)
Comment (alx1313) [reply spec/features/login_spec.rb:35]: Identation checked!

File: spec/login_spec.rb (hunk 12 above)
Comment (CarlosSe) [reply spec/login_spec.rb]: Thanks, we read the information from the link and changed the file to spec / features

File: spec/models/admin_spec.rb (hunk 15 above)
Comment (alx1313) [reply spec/models/admin_spec.rb:1]: Got it!

File: app/models/user.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (alx1313) [reply app/models/user.rb]: whoops, We changed it, now is used to create and login admin instead of users

File: config/environments/development.rb (hunk 5 above)
Comment (alx1313) [reply config/environments/development.rb:40]: We add this to have the functionality of recover password to the admin, but i think probably we will not use it soon, so it will be commented until we need it.
